File: tests/machine/x/zig/tree_sum.zig
const std = @import("std");

const Tree = union(enum) {
    Leaf,
    Node: struct { left: *Tree, value: i32, right: *Tree },
};

const t = Tree{ .Node = .{ .left = &Tree{ .Leaf = {} }, .value = 1, .right = &Tree{ .Node = .{ .left = &Tree{ .Leaf = {} }, .value = 2, .right = &Tree{ .Leaf = {} } } } } };

fn sum_tree(t: *Tree) i32 {
    return switch (t.*) {.Leaf => 0, .Node => |_tmp0| ((sum_tree(&_tmp0.left) + _tmp0.value) + sum_tree(&_tmp0.right)), };
}

pub fn main() void {
    std.debug.print("{d}\n", .{sum_tree(&t)});
}

File: tests/machine/x/dart/tree_sum.dart
abstract class Tree {}
class Leaf extends Tree {
  Leaf();
}
class Node extends Tree {
  Tree left;
  int value;
  Tree right;
  Node(this.left, this.value, this.right);
}

int sum_tree(Tree t) {
  return (() {
  var _t = t;
  if (_t is Leaf) {
    return 0;
  } else if (_t is Node) {
    var left = (_t as Node).left;
    var value = (_t as Node).value;
    var right = (_t as Node).right;
    return (sum_tree(left) + (value as num) as num) + sum_tree(right);
  }  return null;
})();
}

var t = Node(Leaf(), 1, Node(Leaf(), 2, Leaf()));

void main() {
  print(sum_tree(t));
}
